extensions/smallrye-stork/runtime/src/main/java/io/quarkus/stork/StorkConfigUtil.java

```diff
@@ -36,4 +58,218 @@ public static List<ServiceConfig> toStorkServiceConfig(StorkConfiguration storkC
+     * <ol>
+     * <li>If {@code parameters} includes a {@code port}, it is used.</li>
+     * <li>Else, the value from {@code quarkus.http.port} is used, if present.</li>
+     * <li>Otherwise, defaults to {@code 8080}.</li>
+     * </ol>
+     *
+     * @param parameters a map of registrar parameters that may include {@code port}
+     * @param quarkusConfig the active Quarkus {@link Config} instance
+     * @return the resolved port as an integer
+     */
+    public static int getOrDefaultPort(Map<String, String> parameters, Config quarkusConfig) {
+        String customPort = parameters.getOrDefault("port",
+                quarkusConfig.getOptionalValue("quarkus.http.port", String.class).orElse("8080"));
+        return Integer.parseInt(customPort);
+    }
+
+    /**
+     * Attempts to detect the first valid non-loopback IPv4 address of the host machine.
+     * <p>
+     * This is used when no explicit IP address is configured. The method iterates over
+     * all available network interfaces, filtering out loopback and down interfaces.
+     * <p>
+     * If detection fails, it falls back to {@link InetAddress#getLocalHost()}.
+     *
+     * @return the detected {@link InetAddress}, or {@code null} if none found
+     */
+    public static InetAddress detectAddress() {
+        try {
+            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
+            InetAddress addr = findFirstValidAddress(Collections.list(interfaces));
+            if (addr != null) {
+                return addr;
+            }
+        } catch (IOException e) {
+            LOGGER.error("Failed to detect IP address", e);
+        }
+
+        try {
+            return InetAddress.getLocalHost();
+        } catch (UnknownHostException e) {
+            LOGGER.error("Fallback to localhost failed", e);
+            return null;
+        }
+    }
+
+    /**
+     * Finds the first valid IPv4 address from the provided list of network interfaces.
+     * <p>
+     * The method checks that each interface is up and not a loopback, and then looks for
+     * a non-loopback {@link Inet4Address} among the available addresses.
+     *
+     * @param interfaces a list of {@link NetworkInterface} representing system interfaces
+     * @return the first valid {@link InetAddress}, or {@code null} if none found
+     */
+    private static InetAddress findFirstValidAddress(List<NetworkInterface> interfaces) {
+        try {
+            for (NetworkInterface iface : interfaces) {
+                if (!iface.isUp() || iface.isLoopback())
+                    continue;
+
+                Enumeration<InetAddress> addresses = iface.getInetAddresses();
+                while (addresses.hasMoreElements()) {
+                    InetAddress addr = addresses.nextElement();
+                    if (!addr.isLoopbackAddress() && addr instanceof Inet4Address) {
+                        return addr;
+                    }
+                }
```

Comment:
FWIW, this might have a non negligible cost, thus why I'm wondering if we could do better in Docker/Kubernetes to get the IP address?

Reply:
Good point. I thought about optimizing for Docker/K8s, but, as I already mentioned, that’d mean detecting the environment and adding specific logic. 
For now, we keep it generic. Usually, IP and port are configured explicitly, so the auto-detection is mostly a fallback.
In K8s, where fast startups matter, using configured values is always faster and safer.